lib/wrapThenable.js

```diff
@@ -5,42 +5,67 @@ const {
   kReplyHijacked
 } = require('./symbols')
 
-function wrapThenable (thenable, reply) {
+const dc = require('dc-polyfill')
+const channels = dc.tracingChannel('fastify.route')
+
+function wrapThenable (thenable, reply, store) {
   thenable.then(function (payload) {
-    if (reply[kReplyHijacked] === true) {
-      return
+    if (store) {
+      channels.asyncStart.publish(store)
     }
 
-    // this is for async functions that are using reply.send directly
-    //
-    // since wrap-thenable will be called when using reply.send directly
-    // without actual return. the response can be sent already or
-    // the request may be terminated during the reply. in this situation,
-    // it require an extra checking of request.aborted to see whether
-    // the request is killed by client.
-    if (payload !== undefined || (reply.sent === false && reply.raw.headersSent === false && reply.request.raw.aborted === false)) {
-      // we use a try-catch internally to avoid adding a catch to another
-      // promise, increase promise perf by 10%
-      try {
-        reply.send(payload)
-      } catch (err) {
-        reply[kReplyIsError] = true
-        reply.send(err)
+    try {
+      if (reply[kReplyHijacked] === true) {
+        return
+      }
```

Comment:
this should happen before asyncStart, because otherwise there will be no asyncEnd.

Reply:
That's what the try/finally is for. I'm not too familiar with exactly what the purpose is of the `kReplyHijacked` thing there, but I'm going on the assumption that even when the early exit happens we'd still want to see the asyncStart/asyncEnd?

Reply:
you won't get an asyncEnd if hijacked.

Reply:
Why? Wouldn't the try/finally ensure the finally _must_ happen? 🤔